src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -1059,154 +1163,1458 @@ protected EdmTypeKind GetNonEntityValueKind()
+                    isDynamicProperty: null);
+
+                // We don't allow properties or non-custom annotations in the null payload.
+                await this.ValidateNoPropertyInNullPayloadAsync(propertyAndAnnotationCollector)
+                    .ConfigureAwait(false);
+
+                propertyValue = null;
+            }
+            else
+            {
+                string payloadTypeName = null;
+                Tuple<bool, string> readingResourcePropertyResult = await this.ReadingResourcePropertyAsync(
+                    propertyAndAnnotationCollector,
+                    expectedPropertyTypeReference).ConfigureAwait(false);
+                payloadTypeName = readingResourcePropertyResult.Item2;
+
+                if (readingResourcePropertyResult.Item1)
+                {
+                    // Figure out whether we are reading a resource property or not; resource properties are not wrapped while all others are.
+                    // Since we don't have metadata in all cases (open properties), we have to detect the type in some cases.
+                    this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
+
+                    // Now read the property value
+                    propertyValue = await this.ReadNonEntityValueAsync(
+                        payloadTypeName,
+                        expectedPropertyTypeReference,
+                        propertyAndAnnotationCollector,
+                        collectionValidator: null,
+                        validateNullValue: true,
+                        isTopLevelPropertyValue: true,
+                        insideResourceValue: true,
+                        propertyName: null).ConfigureAwait(false);
+                }
+                else
+                {
+                    bool isReordering = this.JsonReader is ReorderingJsonReader;
+
+                    Func<string, Task<object>> readTopLevelPropertyAnnotationValueDelegate =
+                        (annotationName) =>
+                        {
+                            throw new ODataException(ODataErrorStrings.ODataJsonLightPropertyAndValueDeserializer_UnexpectedODataPropertyAnnotation(annotationName));
+                        };
+
+                    // Read through all top-level properties, ignore the ones with reserved names (i.e., reserved
+                    // characters in their name) and throw if we find none or more than one properties without reserved name.
+                    while (this.JsonReader.NodeType == JsonNodeType.Property)
+                    {
+                        await this.ProcessPropertyAsync(
+                            propertyAndAnnotationCollector,
+                            readTopLevelPropertyAnnotationValueDelegate,
+                            async (propertyParsingResult, propertyName) =>
+                            {
+                                if (this.JsonReader.NodeType == JsonNodeType.Property)

```

Comment:
Opportunity to cut down lambda allocations in the future by passing the `JsonReader` as an argument to the lambda.